unixfs/src/walk.rs

```diff
@@ -0,0 +1,1040 @@
+    RootDirectory,
+    /// This is necessarily at the root of the walk
+    BucketAtRoot,
+    /// This is the metadata containing bucket, for which we have a name
+    RootBucket(Cid),
+    /// This is a sibling to a previous named metadata containing bucket
+    Bucket(Cid),
+    /// Directory on any level except root
+    Directory(Cid),
+    /// File optionally on the root level
+    File(Option<Cid>, Option<FileVisit>, u64),
+    /// Symlink optionally on the root level
+    Symlink(Option<Cid>),
+}
+
+/// Representation of the current item of Walker or the last observed item.
+#[derive(Debug)]
+pub enum Entry<'a> {
+    /// Current item is the root directory (HAMTShard or plain Directory).
+    RootDirectory(&'a Path, &'a Metadata),
+    /// Current item is a continuation of a HAMTShard directory. Only the root HAMTShard will have
+    /// file metadata.
+    Bucket(&'a Cid, &'a Path),
+    /// Current item is a non-root plain directory or a HAMTShard root directory.
+    Directory(&'a Cid, &'a Path, &'a Metadata),
+    /// Current item is a possibly root file with a path, metadata, and total file size.
+    File(Option<&'a Cid>, &'a Path, &'a Metadata, u64),
+    /// Current item is a possibly root symlink.
+    Symlink(Option<&'a Cid>, &'a Path, &'a Metadata),
+}
+
+impl<'a> Entry<'a> {
+    /// Returns the path for the latest entry. This is created from UTF8 string and as such always
+    /// representable on all platforms.
+    pub fn path(&self) -> &'a Path {
+        use Entry::*;
+        match self {
+            RootDirectory(p, _)
+            | Bucket(_, p)
+            | Directory(_, p, _)
+            | File(_, p, _, _)
+            | Symlink(_, p, _) => p,
+        }
+    }
+
+    /// Returns the metadata for the latest entry. It exists for initial directory entries, files,
+    /// and symlinks but not continued HamtShards.
+    pub fn metadata(&self) -> Option<&'a Metadata> {
+        use Entry::*;
+        match self {
+            Bucket(_, _) => None,
+            RootDirectory(_, m) | Directory(_, _, m) | File(_, _, m, _) | Symlink(_, _, m) => {
+                Some(m)
+            }
+        }
+    }
+
+    /// Returns the total size of the file this entry represents, or none if not a file.
+    pub fn total_file_size(&self) -> Option<u64> {
+        use Entry::*;
+        match self {
+            File(_, _, _, sz) => Some(*sz),
+            _ => None,
+        }
+    }
+
+    /// Returns the current Cid unelss for root elements.
+    pub fn cid(&self) -> Option<&Cid> {
+        use Entry::*;
+        match self {
+            RootDirectory(_, _) | File(None, _, _, _) | Symlink(None, _, _) => None,
+            Bucket(cid, _)
+            | Directory(cid, _, _)
+            | File(Some(cid), _, _, _)
+            | Symlink(Some(cid), _, _) => Some(cid),
+        }
+    }
+}
+
+impl InnerEntry {
+    fn new_root_dir(_: Cid, metadata: Metadata, name: &str, depth: usize) -> Self {
```

Comment:
is the extra `Cid` param needed in these root-applicable functions?

Reply:
This was the "later" item, "unify something something InnerEntry", I completed that in a recent commit. Resolving.